Primitive aliases (RFC #44)

nmatschke opened the pull request:
Please see [RFC #44](https://github.com/ocaml/RFCs/pull/44) ([rendered version](https://github.com/nmatschke/RFCs/blob/primitive-aliases/rfcs/primitive_aliases.md)) for additional detail.

## Summary

This PR introduces two new forms of primitive declarations, in both structures and signatures:
```ocaml
external foo = Bar.baz
external foo : quux = Bar.baz
```

In either case, the C function names and native argument representations are copied from `Bar.baz` to `foo`. In the first case, the type of `Bar.baz` is copied to `foo`. For the second case, we enforce correctness by unifying `quux` with the type of `Bar.baz`. This means that we do **not** admit the following form:

```ocaml
module _ : sig
  type t
  external int_of_t : t -> int = Fun.id
end = struct
  type t = int
  external int_of_t : t -> int = Fun.id
end
```

This will be addressed in a separate PR, since getting the implementation right in a way that's compatible with recursive modules is tricky.

## Implementation

Our approach is twofold.

First, we refactor the way primitives are represented in the AST, such that they have their own `primitive_description` record type, rather than reusing the same `value_description` as regular OCaml values. The purpose of this is mainly to simplify the rest of the implementation.

Second, we add code to parse the new forms, with distinct representations in the AST. Then, in the typechecker, when we need to translate an alias, we can simply look up the `Longident.t` on the right-hand side in the current environment to find its previously-translated declaration.

nmatschke committed:
Refactor primitive parsing

This commit refactors the way primitives are parsed, in preparation for
the subsequent implementation of primitive aliases (RFC #44). Before
this commit, primitive declarations in both structures and signatures
are parsed into the same `value_description` record as regular OCaml
values. As such, both are represented by the `Psig_value` variant in
signatures. After this commit, primitive declarations are parsed into a
separate `primitive_description` record, and are represented by a
new `Psig_primitive` variant in signatures. In structures,
`Pstr_primitive` is  updated to contain a `primitive_description` rather
than a `value_description`.

We also add a new `Pstr_val` node for the special case of `val v : t`
syntax in structures, which is supported for ppxes, but rejected by the
typechecker.

Finally, we propagate this split to the typed tree as well, resulting in
separate `Tsig_value of value_description` and
`Tsig_primitive of primitive_description` nodes.

There are some additional changes to tests which print out
representations of the AST.

lpw25 commented:
>  we enforce correctness in structures by unifying quux with the type of Bar.baz, but only enforce correctness in signatures via the inclusion check

I think this might cause some subtle issues around recursive modules. In particular, I worry that your PR might accept:
```ocaml
external foo : int -> int = "%identity"
module rec M : sig external bar : int -> string = foo end = M
```
Generally we insist that signatures be sound on their own terms to avoid such issues -- which would mean ruling out the example you gave in the description. It's possible you could instead catch this case by trying to unify the types in signatures, and if that fails then annotating the `external` with this fact, than in `Translmod.init_shape` such an external would raise an error.

nmatschke commented:
@lpw25 it does indeed accept that. Will think more about how to fix it.

nmatschke commented:
@lpw25 I have implemented the fix as discussed (offline) and added a couple test cases to recursive_modules.ml in the testsuite.

yallop commented:
> For the second case, we enforce correctness in structures by unifying `quux` with the type of `Bar.baz`,

The current implementation doesn't seem quite right: it types

```ocaml
external foo : int -> 'a = "%foo"
external bar : 'a -> 'a = foo
```

as

```ocaml
external foo : int -> int = "%foo"
external bar : int -> int = "%foo"
```

nmatschke commented:
@yallop I'm happy to modify the behavior to something else, but I'd argue that this is consistent with how `let` bindings work in structures. Note that your example would be rejected by the module inclusion check, if given a signature:
```ocaml
# module _ : sig 
    external foo : int -> 'a = "%identity"
    external bar : 'a -> 'a = foo
  end = struct
    external foo : int -> 'a = "%identity"
    external bar : 'a -> 'a = foo
  end
  ;;
```
```
Error: Signature mismatch:
       Modules do not match:
         sig
           external foo : int -> int = "%identity"
           external bar : int -> int = "%identity"
         end
       is not included in
         sig
           external foo : int -> 'a = "%identity"
           external bar : 'a -> 'a = "%identity"
         end
       Values do not match:
         external foo : int -> int = "%identity"
       is not included in
         external foo : int -> 'a = "%identity"
       The type int -> int is not compatible with the type int -> 'a
       Type int is not compatible with type 'a
```

yallop commented:
It's not quite consistent with the behaviour for `let` bindings: with `let` bindings in structures the type of `foo` wouldn't be updated by the declaration of `bar`.

The module inclusion check does indeed do what I'd expect, but I find the behaviour of primitive aliases signatures surprising, as well.  Currently, the following is accepted:

```ocaml
module type s = sig
    external foo : int -> int = "%identity"
    external bar : float = foo
  end                        
```

even though there's no relationship at all between the type of `foo` and the type of `bar`.

My personal preference would be to have aliases in signatures and in structures operate consistently, with
- type variables treated as rigid (not unifiable metavariables) in both cases
- `external f : t = g` valid only if `t` is an _instance_ of the type of `g`

lpw25 commented:
I agree that having `bar` update `foo`'s type is weird. I think that `external` declarations without aliases should definitely be treated as rigid.

I think that supporting things like:
```ocaml
external bar : _ -> _ = foo
```
is reasonable and suggests treating type annotations on aliases as not rigid.

I see how supporting this example:
```ocaml
module _ : sig
  type t
  external int_of_t : t -> int = Fun.id
end = struct
  type t = int
  external int_of_t : t -> int = Fun.id
end
```
naturally leads to accepting:
```ocaml
 module type s = sig
    external foo : int -> int = "%identity"
    external bar : float = foo
end
```
but I wonder if we can't use `Ctype.mcomp` to at least have it error if there is no way the types could be made equal.

lpw25 commented:
> I wonder if we can't use Ctype.mcomp to at least have it error if there is no way the types could be made equal.

In fact this should be the same as doing the unification with the GADT index rules (i.e. "pattern mode") rather than the normal rules ("expression mode").

yallop commented:
> I think that supporting things like:
> 
> ```ocaml
> external bar : _ -> _ = foo
> ```
>
> is reasonable 

One alternative way of supporting this pattern would be to allow

```ocaml
external bar = foo
```

Are there any useful cases that wouldn't be covered with this approach?

[Edit: in fact, that syntax is already allowed with the implementation in this PR!]

yallop commented:
Answering my own question, the underscore is a useful way to specialize only part of a signature:

```ocaml
external length : 'a array -> int = "%length" 
external ilength : int array -> _ = length
```

nmatschke commented:
Sorry, it took a while to get back to this. I've implemented the suggested change, to instead use `matches` to check whether the primitive's type could be instantiated to that of its alias, and using the GADT unification logic in signatures, where we might not know about certain type equations. Tests reflect the new behavior.

Note that this is rejected:
```ocaml
external length : 'a array -> int = "%array_length"
external ilength : int array -> _ = length
```
With:
```
Error: The type of this alias does not match that of the aliased primitive.
       Type int array -> 'b is not compatible with type 'a array -> int
```
I think that's not unreasonable behavior, though.

nmatschke commented:
@Octachron could you take a look at this PR, please, at your convenience? Thanks!

Octachron commented:
Thanks for the request ! This PR does sound like a good fit for committee consideration, I would keep you updated once the committee has chosen a shepherd.

Octachron commented:
@yallop has accepted to shepherd the pull request, in other words, to summarize it to the language committee and propose a recommendation for the committee to debate.

yallop commented:
I'm currently working on clarifying my understanding of the proposal, in the hope of proposing acceptance to the Committee. 

A few questions to start off with:

1. I think the issues around unification discussed above are resolved now.  To put it in high-level terms, I think the types in structure items are now handled similarly to the signature inclusion check.  That is:

   ```ocaml
   external p : t = "x"
   external q : s = p
   ```

   is acceptable when the signature ascription

   ```ocaml
   struct let p : t = ... end
    : sig val p : s end
   ```

    is acceptable.  Is that understanding correct?

2. I'm uneasy about the permissiveness of checking for aliases in signatures.  Delaying checking until signature inclusion means allowing obviously incorrect declarations, e.g.

   ```ocaml
   module type S = sig external f : int = Fun.id end
   ```

   Would it be possible to use the compatibility check on aliases in signatures to exclude these kinds of aliases that can never be instantiated?

3. Is the intent of a primitive alias that the attributes of the original binding should be exactly carried over to the new definition (i.e. that attributes cannot be added or dropped)?

nmatschke commented:
Thanks @yallop.

1. Yes, that essentially matches my understanding. Pedantically, I think it is acceptable when the signature ascription `struct external p : t = ... end : sig val p : s end` is acceptable, since the type variables in `t` and `s` are both generic. I.e. e.g. `struct external p : 'a -> 'a = "%identity" end : sig val p : 'a -> 'a end` is acceptable but `struct let p : 'a -> 'a = ( ~- ) end : sig val p : 'a -> 'a end` is not.
2. I'm not sure it's worth trying to prevent such constructions. For one, this is already allowed in OCaml today: `module type S = sig external f : int = "%identity" end`. Also, if one were to (obviously unsafely) produce an `(int, string -> string) Type.eq`, they could implement such an interface. My understanding is that the latter is essentially what using `unify_gadt` tells us, and I don't immediately see an easy fix that also preserves its desirable properties.
3. Yes, the intent is that the attributes of the original binding should be carried over exactly as-is. Attributes on the alias would have no effect; only the type matters.

yallop commented:
Thank you for the explanations, @nmatschke.  The first point (unification) seems fine to me now.

The third point (attributes on aliases) also seems mostly fine, although if attributes on the alias have no effect then I'd think that we ought to at least issue a warning about them.

On the second point (compatibility), I don't entirely follow your explanation, but the scenario involving the GADT doesn't seem realistic.  In contrast, a user inadvertently writing an incompatible type in an primitive alias seems quite likely, so it'd be useful to know whether  a compatibility check is feasible.  Perhaps @lpw25 could offer a view.

Octachron commented:
The `int` case sounds like a bug (the nominal nature of the `int` type is lost somewhere), a variation with an explicitly distinguishable type
```ocaml
type u = A
module type S = sig external x: u = Fun.id end
```
fails as expected with
```
Error: The type of this alias does not match that of the aliased primitive.
       Type u is not compatible with type 'a -> 'a
```

lpw25 commented on typing/ctype.ml:
```diff
@@ -2297,9 +2297,11 @@ let find_expansion_scope env path =
+let non_aliasable env p decl =
   (* in_pervasives p ||  (subsumed by in_current_module) *)
-  in_current_module p && not decl.type_is_newtype
+  not (Env.is_in_signature env) &&
```
I think this needs to be:
```
in_pervasives p
|| (not (Env.is_in_signature env)
   && in_current_module p
   && not decl.type_is_newtype)
```

lpw25 commented:
> I'm uneasy about the permissiveness of checking for aliases in signatures. Delaying checking until signature inclusion means allowing obviously incorrect declarations, e.g.
> ```ocaml
> module type S = sig external f : int = Fun.id end
> ```
> Would it be possible to use the compatibility check on aliases in signatures to exclude these kinds of aliases that can never be instantiated?

The current implementation prevents writing declarations that can never possibly be instantiated. It happened to have a bug which meant this didn't work correctly for `int` and other types in the initial environment.

I think this is probably a reasonable compromise between allowing the flexibility to re-export externals for abstract types and giving early errors to users for obvious mistakes.

yallop commented:
I agree.  My concerns have been addressed, and I'm planning to write up a recommendation for the committee soon.

gasche commented:
Question: a natural extension of the language would be to allow `let`-binding in signatures, restricted to a name or a value.

```ocaml
module _ : sig
  type t
  let int_of_t : t -> t = Fun.id
end
```

This feature would make sense for arbitrary identifiers, not just those declared as external primitives.

Why is it important to restrict rebinding to external primitives?

SkySkimmer commented:
>Question: a natural extension of the language would be to allow let-binding in signatures, restricted to a name or a value.

Reusing `let` opens questions we may not want to bother with. For instance, what is the module type of
~~~ocaml
struct
  let x = y
end
~~~
? To make let in sig work I guess it would need to be `sig let x = y end` but I don't think that's backward compatible.

Is
~~~ocaml
struct
  let y = z
  let x = y
end : sig let x = z end
~~~
accepted?

etc

nmatschke commented:
@yallop 
> The third point (attributes on aliases) also seems mostly fine, although if attributes on the alias have no effect then I'd think that we ought to at least issue a warning about them.

The compiler will already emit misplaced-attribute warnings for built-in attributes on primitive aliases. I've [added some examples](https://github.com/ocaml/ocaml/pull/13377/commits/0c6da5d4aec6a72537ceadabc9f413944dedbeac) to the existing w53 tests demonstrating this.

@lpw25 
> The current implementation prevents writing declarations that can never possibly be instantiated. It happened to have a bug which meant this didn't work correctly for int and other types in the initial environment.

I fixed this via the [change you suggested](https://github.com/ocaml/ocaml/pull/13377#discussion_r1993908017) and [added tests](https://github.com/ocaml/ocaml/pull/13377/commits/1354af325c8541a6f7fe10aa0da372bf95e46146).

@gasche 
> Why is it important to restrict rebinding to external primitives?

I suppose it is a bit weird that primitive aliases are exposed in signatures. However, without this feature, I'd argue that external primitives are considerably less ergonomic than value declarations, as one must include the name of the C function(s) and any relevant attributes, in addition to the type. Also, frankly, the implementation of such a feature for `let` bindings seems more complex, requiring a path equality check in module inclusion, whereas we essentially get primitive aliases in signatures "for free" by virtue of reusing `Typedecl.transl_prim_desc`, and can simply expand the aliases and rely on the existing inclusion check logic. Perhaps at some point in the future we can add such a feature.

yallop commented:
I'm pleased to report that the Committee reached consensus in favour of this design.  Practically speaking, that means that once the implementation has been reviewed we can merge it without further ado.

nmatschke commented:
Awesome! By the way, @yallop, the CI is complaining about the "Parsetree Updated: comment-and-label" job. My impression is that this requires adding a label to the PR, which I'm not sure I have permission to do. Could you take a look at this, please?

yallop commented on lambda/translmod.ml:
```diff
@@ -258,8 +259,13 @@ let init_shape id modl =
+        | Unsafe_in_recmod ->
+          raise (Initialization_failure
+                  (Unsafe {reason=Unsafe_primitive; loc; subid}))
+        end
```
```suggestion
    | Sig_value(_, {val_kind=Val_prim (_, Safe)}, _) :: rem ->
       init_shape_struct env rem
    | Sig_value(subid, {val_kind=Val_prim (_, Unsafe_in_recmod); val_loc = loc}, _) :: _ ->
       raise (Initialization_failure (Unsafe {reason=Unsafe_primitive; loc; subid}))
```

yallop commented on manual/src/refman/extensions/recursivemodules.etex:
```diff
@@ -54,9 +54,11 @@ not final and subject to change in future releases.
+not contain any primitive aliases with more-general type than that of
+the aliased primitive.  Evaluation of a recursive module definition
+proceeds by building initial values for the safe modules involved, binding
+all (functional) values to
```
I suggest phrasing this positively to match the previous clause ("if all value definitions ... if every primitive alias ..."):

```suggestion
it contains have function types @typexpr_1 '->' typexpr_2@, and if every
primitive alias that it contains has a type no more general than that of
the aliased primitive.  Evaluation of a recursive module definition
proceeds by building initial values for the safe modules involved, binding
all (functional) values to
```

yallop commented on typing/untypeast.ml:
```diff
@@ -22,6 +22,8 @@ open Ast_helper
 module T = Typedtree
 
+type ('a, 'b) either = Left of 'a | Right of 'b
+
```
We have `Either.t` in the standard library, so there's no need to define it here

```suggestion
```

yallop commented on typing/untypeast.ml:
```diff
@@ -69,7 +71,8 @@ type mapper = {
   value_binding: mapper -> T.value_binding -> value_binding;
-  value_description: mapper -> T.value_description -> value_description;
+  value_description: mapper -> Typedtree.value_description
+    -> (value_description, primitive_description) either;
```
```suggestion
    -> (value_description, primitive_description) Either.t;
```

yallop commented on typing/untypeast.ml:
```diff
@@ -197,10 +203,23 @@ let structure_item sub item =
-    (sub.typ sub v.val_desc)
+  match v.val_ext with
+  | Tval_val typ ->
+    Left
```
```suggestion
    Either.Left
```

yallop commented on typing/untypeast.ml:
```diff
@@ -197,10 +203,23 @@ let structure_item sub item =
+        (map_loc sub v.val_name)
+        (sub.typ sub typ))
+  | Tval_prim_decl (typ, prim) ->
+    Right
```
```suggestion
    Either.Right
```

yallop commented on typing/untypeast.ml:
```diff
@@ -197,10 +203,23 @@ let structure_item sub item =
+        (map_loc sub v.val_name)
+        (sub.typ sub typ))
+  | Tval_prim_alias (typ, lid) ->
+    Right
```
```suggestion
    Either.Right
```

yallop commented on typing/untypeast.mli:
```diff
@@ -17,6 +17,8 @@ open Parsetree
 val lident_of_path : Path.t -> Longident.t
 
+type ('a, 'b) either = Left of 'a | Right of 'b
+
```
```suggestion
```

Octachron commented on manual/src/cmds/intf-c.etex:
```diff
@@ -48,6 +49,12 @@ in library modules if they have side-effects at toplevel, as this
+function name(s), and any additional information as discussed in
+section~\ref{s:C-cheaper-call}. One may optionally ascribe a type to the new
+binding as well.
+
```
You should mention that this is only possible starting with `OCaml 5.x` .

nmatschke replied:
Done.

Octachron replied:
This will be 5.6 rather than 5.5 .

Octachron commented on parsing/ast_iterator.ml:
```diff
@@ -616,13 +618,22 @@ let default_iterator =
+        (match pprim_kind with
+         | Pprim_decl (pprim_type, _) -> this.typ this pprim_type
+         | Pprim_alias (pprim_type, _) ->
+           Option.iter (this.typ this) pprim_type);
```
There is a missing call to `iter_loc_lid`.

yallop commented:
> Awesome! By the way, @yallop, the CI is complaining about the "Parsetree Updated: comment-and-label" job. My impression is that this requires adding a label to the PR, which I'm not sure I have permission to do. Could you take a look at this, please?

@Octachron has added the label, but there's some kind of [permissions problem](https://github.com/ocaml/ocaml/actions/runs/14412088239/job/40424083488?pr=13377#step:2:50) that I don't know how to fix.  Perhaps the CI configuration needs to be updated.

yallop commented on typing/typedecl.ml:
```diff
@@ -1629,63 +1630,147 @@ let check_unboxable env loc ty =
+                         Primitive_alias_unification (env, err)))
+           in
+           let val_kind =
+             match
+               Ctype.matches ~expand_error_trace:true env
+                 cty.ctyp_type v.val_type
+             with
+             | () -> val_kind
+             | exception (Ctype.Matches_failure _)
+               when Env.is_in_signature env ->
+               (match
+                  Ctype.matches_gadt ~expand_error_trace:true env
+                    cty.ctyp_type v.val_type
+                with
+                | () -> Val_prim (desc, Unsafe_in_recmod)
+                | exception (Ctype.Matches_failure (env, err)) ->
+                  raise_unification_error env err)
+             | exception (Ctype.Matches_failure (env, err)) ->
+               raise_unification_error env err
```
The control flow here (and perhaps also the logic) is rather complicated.  If there's a way to simplify it, that'd be good, but I'd appreciate at least seeing a comment explaining the aim of using `Ctype.matches`, `Ctype.matches_gadt` in this combination.

nmatschke replied:
I pushed some comments explaining this code.

Octachron commented on parsing/ast_mapper.ml:
```diff
@@ -688,14 +691,30 @@ let default_mapper =
+           Prim.mk_alias
+             (map_loc this pprim_name)
+             (Option.map (this.typ this) pprim_type)
+             pprim_ident
```
The ident should be mapped with `map_loc_lid`.

Octachron commented on parsing/depend.ml:
```diff
@@ -388,6 +388,12 @@ and add_sig_item (bv, m) item =
+      (match pd.pprim_kind with
+      | Pprim_decl (pprim_type, _) | Pprim_alias (Some pprim_type, _) ->
+        add_type bv pprim_type
+      | Pprim_alias (None, _) -> ());
```
The `longident` part of the alias can contain dependencies too.

Octachron commented on parsing/depend.ml:
```diff
@@ -526,8 +532,12 @@ and add_struct_item (bv, m) item : _ String.Map.t * _ String.Map.t =
+      (match pd.pprim_kind with
+       | Pprim_decl (pprim_type, _) | Pprim_alias (Some pprim_type, _) ->
+         add_type bv pprim_type
+       | Pprim_alias (None, _) -> ());
```
Sale as `add_sig_item`: there is a missing call to `add bv aliased_name`.

Octachron commented on parsing/parser.mly:
```diff
@@ -1532,10 +1537,11 @@ structure_item:
     | value_description
-        { pstr_primitive $1 }
+        { ill_formed_ast $sloc
+          "Value declarations are only allowed in signatures" }
```
This syntax is explicitly supported for ppxs: #6688 and there are discussion to allow it in the surface language in https://github.com/ocaml/RFCs/pull/59 . Thus I think it might make sense to create a new AST node for this syntax to avoid the awkward situation where a syntactic form becomes temporarily absent from the language and then reappear later.

nmatschke replied:
I reverted this change and added a `Pstr_val` AST node.

Octachron commented on parsing/parsetree.mli:
```diff
@@ -1067,9 +1088,8 @@ and structure_item_desc =
-      (** - [val x: T]
-            - [external x: T = "s1" ... "sn" ]*)
+  | Pstr_primitive of primitive_description
+      (** [external x: T = "s1" ... "sn" ] *)
```
The ppx syntax should still be supported. You probably also want to add primitive aliases to the description.

nmatschke replied:
Done.

Octachron commented on parsing/parsetree.mli:
```diff
@@ -917,9 +938,9 @@ and signature_item =
-         *)
+      (** [val x: T] *)
+  | Psig_primitive of primitive_description
+      (** [external x: T = "s1" ... "sn"] *)
```
You are missing the new syntax 
```ocaml
external x : T = y
```
in the description.

Also in general, I am not fond of splitting the signature item between `Psig_value _` and `Psig_primitive _` only at the parsetree level while keeping only one item at the typedtree level. 

I would propose to choose one option and use it uniformly in `Parsetree` and `Typedtree` (we want to keep one constructor at the type level).

In term of choice, splitting the constructor has the advantage of making it clearer that there are no mutually recursive primitives whereas keeping a shared constructor emphasizes that in both cases we are defining a new name at the value level; and it make it easier to support
```ocaml
val x: int
```
at the structure level for now.

Personally, I tend to agree with your choice of splitting the signature item in two.

nmatschke replied:
How's this? I opted to propagate the split into typedtree rather than remove the split in parsetree.

Octachron commented on typing/tast_iterator.ml:
```diff
@@ -169,7 +169,10 @@ let value_description sub x =
+  match x.val_ext with
+  | Tval_val typ | Tval_prim_decl (typ, _) | Tval_prim_alias (Some typ, _) ->
+    sub.typ sub typ
+  | Tval_prim_alias (None, _) -> ()
```
There is is missing call to `iter_loc_lid` on the aliased primitive.

Octachron commented on typing/tast_mapper.ml:
```diff
@@ -193,9 +193,15 @@ let structure_item sub {str_loc; str_desc; str_env} =
+    | Tval_val typ -> Tval_val (sub.typ sub typ)
+    | Tval_prim_decl (typ, prim) -> Tval_prim_decl (sub.typ sub typ, prim)
+    | Tval_prim_alias (typ, lid) ->
+      Tval_prim_alias (Option.map (sub.typ sub) typ, lid)
```
The aliased `lid` should be mapped with `map_loc_lid`.

Octachron commented on typing/typedecl.ml:
```diff
@@ -2331,6 +2414,19 @@ let report_error ~loc = function
+  | Primitive_alias_does_not_refer_to_primitive kind ->
+      Location.errorf ~loc
+        "@[This@ identifier@ should@ be@ a@ primitive,@ but@ it@ is@ bound@ \
+         to@ %s.@]"
```
Side-note: opening a box here is problematic because it fixes the left-margin. But it seems fine to keep it here, since I will have to fix the other error messages later on.

Octachron commented on typing/typedecl.ml:
```diff
@@ -2331,6 +2414,19 @@ let report_error ~loc = function
+      let msg = Format_doc.Doc.msg in
+      Location.errorf ~loc
+        "@[<v>The type of this alias does not match that of the aliased \
+         primitive.@\n%t@]" @@ fun ppf ->
```
Inside a `v` box, you don't need to use `@\n`: using `@,` is better (the semantics of `@\n` are a bit overcomplicated currently).

nmatschke replied:
Like this?

Octachron commented on typing/typedecl.ml:
```diff
@@ -75,6 +74,8 @@ type error =
   | Nonrec_gadt
   | Invalid_private_row_declaration of type_expr
+  | Primitive_alias_does_not_refer_to_primitive of string
+  | Primitive_alias_unification of Env.t * Errortrace.unification_error
```
I would suggest `Primitive_type_mismatch` in order to have a name less context-dependent.

Octachron commented:
The permission problem can be ignored (the CI test needs to be simplified).

There is still an issue with recursive modules, the following definition is accepted

```ocaml
type e
external id: e -> e = "%identity"
type f = int -> int


module rec A: sig
  type t
  val a: t
  external id: A.t -> f = id
end = struct
  type t = A
  external id = A.id
  let a = A
end
```
because we trust the signature-side type of the alias in the implementation.

Octachron commented:
I am inclined to propose to split the PR in two parts 

- A first part for primitive aliases with matching type annotations
- An extension for handling primitive aliases with not-matching-but-possibly-compatible type annotations in signature

to allow faster review and integration (of both parts ultimately).

Indeed, my impression after a first review is that the first part is quite close to be ready to be merged: there are some issues remaining with signature avoidance and other minor points, but nothing major.
Contrarily, the extension requires more thoughts about the interaction with recursive modules but the related changes are probably quite localized.

Thus by splitting the PR in two parts I hope that we will be able to quickly merge the first part, and then move to a more focused discussion for the extension.

What is your opinion @nmatschke on this potential split?

nmatschke commented:
Thanks for the review @yallop. I pushed some changes that (I think) resolve the issues you raised.

@Octachron ah yeah, ~~I see how it's tricky to get the `non_aliasable` check to have the right behavior here~~ actually I think this might be an issue with how recursive module safety is implemented. I don't mind splitting the PR up. Unless you object, I will remove everything related to recursive modules and `matches_gadt` from this PR, and move it to a second PR.

Octachron commented:
@nmatschke , I agree that splitting the `matches_gadt` part and the related recursive modules check in a second PR is probably a good way to make progress on the PR, sorry for the late reply.

nmatschke commented:
Sorry for letting this sit for so long! I just pushed some changes backing out the `matches_gadt` and recursive module code. I opened a second PR (#14566) for that discussion.

I also squashed a bunch of commits to produce a clean history.

Octachron commented on typing/typedecl.ml:
```diff
@@ -81,6 +80,8 @@ type error =
   | Invalid_private_row_declaration of type_expr
   | Atomic_field_must_be_mutable of string
   | External_with_non_syntactic_arity
+  | Primitive_alias_does_not_refer_to_primitive of string
```
It is better to keep the `value_kind` in the error type, and do the translation from the error type to a human-readable error message in one place. With a `string`, the typechecking code end up doing part of the translation with a partial context.

nmatschke replied:
Done.

Octachron commented on typing/typemod.ml:
```diff
@@ -1140,6 +1140,22 @@ and approx_sig env ssg =
+              Sig_value(id, v, Exported) :: approx_sig env srem
+            | exception Not_found -> approx_sig env srem
+            end
+          end
```
This seems to be a remnant from the support of aliasing through recursive module definition. I think it is better to remove it in the present PR.

Indeed, in the approximation phase we are only interested into building a skeleton of the module type, primitive aliases without type annotations mixes the value level and type level which brings new complication that we want to avoid for now. For instance, the current code fails if the fix point is not reached in two steps:
```ocaml
module rec X1 : sig
  external p = Obj.magic
end = struct external p = Obj.magic end

and X2 : sig
  external p = X1.p
end = struct external p = X1.p end

and X3 : sig
  external p = X2.p
end = struct external p = X2.p end
```

>```
>14 |   external p = X2.p
>       ^^^^^^^^^^^^^^^^^
>Error: Unbound value "X2.p"
>```

nmatschke replied:
Hmm, even with some of the other code I've written to support recursive modules, you're right that this doesn't quite have the right behavior. Deleted the `Psig_primitive` case.

I think it's unfortunate that this results in errors like
```
Error: Unbound value "E.p"
```
as opposed to
```
Error: This module type is recursive. This use of the recursive module "E"
       within the definition of the module "D"
       makes the module type of "D" depend on the module type of "E".
       Such recursive definitions of module types are not allowed.
```
But to get the latter, we'd need to call `Env.lookup_value` or similar here. Do you think that's worth it?

We could always try `Typedecl.transl_prim_desc`, but I think this is actually worse for supporting recursive modules, because it calls `Typetexp.transl_type_scheme` and we don't have types from the current recursive module in the environment. So I think we'd really need a way to approximate primitives to improve here, apart from calling `Env.lookup_value` just to get a better error message.

Octachron commented on typing/typedtree.mli:
```diff
@@ -714,12 +714,16 @@ and object_field_desc =
-    val_desc: core_type;
     val_val: Types.value_description;
-    val_prim: string list;
+    val_ext: value_externality;
```
My feeling is that the `val_ext: value_externality`names do not reflect well the normal case. Maybe a name like `val_kind: value_kind` would better include this case.

nmatschke replied:
Done.

Octachron commented on typing/typedecl.ml:
```diff
@@ -1653,63 +1654,139 @@ let check_unboxable env loc ty =
+             (* When the alias has a type ascription, we check that it is no
+                more general than the type of the aliased declaration. This
+                check is conservative in signatures, as any structure may reveal
+                additional equalities which would make the types match. *)
```
This comment is out-of-sync with the current code which no longer tries with `Ctype.gadt_matches` upon a `Ctype.matches` failure.

nmatschke replied:
Just not mentioning anything about the differences between structures and signatures seems alright to me.

Octachron reviewed (commented):
I think there are still two small designs question to resolve about the parsetree signature items

- Do we want to split the `Pstr_value` nodes in two in both the `Parsetree` and the `Typedtree`?
- How do we keep support for `struct val x: int end` (which is allowed in ppxs) ?

Otherwise, beyond a handful of remarks (in particular about remnant of the recursive module), the code seems in good shape.

Octachron commented on parsing/ast_helper.mli:
```diff
@@ -317,7 +328,8 @@ module Str:
     val eval: ?loc:loc -> ?attrs:attributes -> expression -> structure_item
     val value: ?loc:loc -> rec_flag -> value_binding list -> structure_item
-    val primitive: ?loc:loc -> value_description -> structure_item
+    val val_: ?loc:loc -> value_description -> structure_item
```
Self-note: The Pstr_val parsetree nodes are part of the extended syntax, whereas the Ast_helper functions aim more to build vanilla syntax nodes. Thus it might make sense to not expose this function. But it is probably more uniform to define it. (So you can ignore this comment).

Octachron commented on parsing/pprintast.ml:
```diff
@@ -1074,11 +1074,22 @@ and floating_attribute ctxt f a =
+
+and primitive_description ctxt f x =
+  (* note: primitive_description has an attribute field,
+           but they're already printed by the callers this method *)
```
`by the callers this method` ⇒ `by the callers of this method`

Octachron commented on typing/typedtree.mli:
```diff
@@ -716,10 +717,22 @@ and value_description =
+
+and primitive_kind =
+  | Tprim_decl of core_type * string list
+  | Tprim_alias of core_type option * Longident.t loc
```
The argument of the `Tprim_alias` constructor is missing the `Path.t` corresponding to the `Longident.t`: the typedtree must contain all name resolutions.

nmatschke replied:
Done.

Octachron commented on file_formats/cmt_format.ml:
```diff
@@ -330,10 +331,10 @@ let iter_on_occurrences
-      | Tsig_module _ | Tsig_recmodule _ | Tsig_modtype _ | Tsig_modtypesubst _
-      | Tsig_open _ | Tsig_include _ | Tsig_class _ | Tsig_class_type _
-      | Tsig_attribute _ -> ());
+      | Tsig_value _ | Tsig_primitive _ | Tsig_type _ | Tsig_typesubst _
```
In the alias case, the aliased term must be recorded (like in the `Tsig_modsubst` case): 
if someone rename `id` to `newid` in
```ocaml
external id: int -> int = "%identity"
external id2 = id
````
the occurrence of `id` in the primitive `id2` should be replaced by `newid` too.

nmatschke replied:
Look good?

Octachron replied:
Indeed, thanks !

Octachron commented on typing/typedecl.ml:
```diff
@@ -1653,63 +1654,132 @@ let check_unboxable env loc ty =
 let transl_value_decl env loc valdecl =
   let cty = Typetexp.transl_type_scheme env valdecl.pval_type in
   let ty = cty.ctyp_type in
+  assert (Env.is_in_signature env);
```
The confusion is not really possible anymore. I would propose to remove this assert.

Octachron commented on typing/typedecl.ml:
```diff
@@ -1653,63 +1654,132 @@ let check_unboxable env loc ty =
+    in
+    let val_kind = v.val_kind in
+    (match val_kind with
+     | Val_prim _ ->
```
Possible alternative to make it clearer that we are erroring outside of the `Val_prim` case
```ocaml
    begin match v.val_kind with
    | Val_prim _ -> ()
    | Val_reg | Val_ivar _ | Val_self _ | Val_anc _ as k ->
        raise(Error(pprim_ident.loc,
                    Primitive_alias_does_not_refer_to_primitive k))
    end;
````

nmatschke replied:
I'm not sure I totally follow your point here. I feel like the current structure is pretty clear, especially after removing some of the unnecessary complexity around `val_kind`, and I'd prefer not to match on it twice.

Octachron replied:
My (sibylline) point was that the code is currently only using the pattern matching to raise an error in the error branch, but it is not using the `Val_prim _ ` pattern in the main branch, since you are not binding anything in this branch.

Thus an alternative way to write
```ocaml
match v.val_kind with
| Val_prim _ -> $body
| Val_reg | Val_ivar _ | Val_self _ | Val_anc _ as val_kind ->$error_body
```
would be
```ocaml
 begin match v.val_kind with
| Val_reg | Val_ivar _ | Val_self _ | Val_anc _ as val_kind ->$error_body
| Val_prim _ -> ()
end;
$body
```
which emphasizes more that the first part is only raising an error if the condition `v.val_kind = Val_prim _ ` is false.

Another option would be to commute the two patterns
```ocaml
match v.val_kind with
| Val_reg | Val_ivar _ | Val_self _ | Val_anc _ as val_kind ->$error_body
| Val_prim _ -> $body
```
to let the reader knows immediately that the other cases raise an error.

But this was more a comment on possible alternative and equivalent ways to present the code, not a request for changes.

nmatschke replied:
Makes sense, thanks. Perhaps unsurprisingly, I slightly prefer the way I wrote it, but don't feel all that strongly.

If I had to make an argument, there are two components:
- I'd guess the reader is more likely to be interested in the happy case than the error case, so putting the error case first is visual noise to filter out most of the time. (I could easily be wrong about this.)
- I'd still like to merge one of #14566 or #14636, in which the happy case needs some of the information from the `Val_prim` constructor. This is what I was referring to by "match on it twice"; sorry for getting ahead of myself.

Octachron commented on typing/typedecl.ml:
```diff
@@ -1653,63 +1654,132 @@ let check_unboxable env loc ty =
+               raise(Error(cty.ctyp_loc, Primitive_type_mismatch (env, err)))
+           in
+           Some cty,
+           { v with val_type = cty.ctyp_type; val_kind; val_loc = loc }
```
My feeling is that this pattern hides a bit that if we reach this point `val_kind` is `v.val_kind`.

nmatschke replied:
Good catch, this structure was a holdover from dealing with recursive modules. Is this better?

Octachron replied:
The new code does look better to me.

Octachron commented on typing/printtyped.ml:
```diff
@@ -490,8 +490,19 @@ and value_description i ppf x =
+    list (i+1) string ppf prim
+  | Tprim_alias (typ, lid) ->
+    option (i+1) core_type ppf typ;
+    longident (i+1) ppf lid
```
Here, it should be the missing path which is printed.

nmatschke replied:
Look good?

Octachron replied:
Yes, the printing matches more the contents of the typedtree.

Octachron reviewed (commented):
The PR looks quite close to be ready to merge. 

There is one point left to fix: the typedtree primitive description does not contain the resolved path for alias. Consequently, this lack of path implies that the occurrence index in cmt files are missing the aliased primitives.

Also the manual should be updated in the extension sections of the reference manual but that can be done later.

(I have also some minor style remarks that can be ignored if you wish)

Octachron commented on parsing/parser.mly:
```diff
@@ -3261,7 +3265,7 @@ primitive_declaration:
       let loc = make_loc $sloc in
       let docs = symbol_docs $sloc in
-      Val.mk id ty ~prim ~attrs ~loc ~docs,
+      Prim.mk id ty ~prim ~attrs ~loc ~docs,
```
My impression is that in isolation those refactorization changes in `parser.mly` are not enough to justify regenerating the whole parser. I propose that you move those changes to the next commit while keeping the refactorisation of the parsetree in this first commit.

Octachron commented:
Looking at the bootstrap in the two commits, there are not necessarily currently since primitive aliases are not used in the compiler and the changes do not affect any of the produced artefacts that are shared between multiple stages.

Thus I propose that you remove the bootstrapping changes from the two commits, it order to reduce the size of the committed bytecode in the repo.

Octachron reviewed (approved):
Outside of the bootstrapping minutiæ for the parser and ocamlc, I believe that the pull request is ready to be merged.

nmatschke committed:
Implement primitive aliases (RFC #44)

This commit implements primitive aliases (RFC #44). Explicitly, we parse
two new forms in both structures and signatures:
- `external p = M.x`
- `external p : t = M.x`

In both cases, the C function names and native representation of
arguments are copied from `M.x`, rather than requiring the user to
repeat them in full.

In the first case, the type is copied from `M.x`.

In the second case, we assert that the supplied type matches that of
`M.x`.

nmatschke commented:
@Octachron I think I addressed your points about the parser and ocamlc.

Octachron commented:
Thanks the PR is indeed ready to merge (except for the parser conflict, do you want me to take care of it?).

nmatschke commented:
Fixed the parser. I don't think I can merge the PR myself, so if you could take care of whatever is needed to make that happen, it would be much appreciated. Thanks!

Octachron commented:
Merged, thanks for your patience!
